xcommon/src/llvm.rs

```diff
@@ -0,0 +1,107 @@
+//! LLVM utilities
+
+use anyhow::{bail, ensure, Result};
+use std::collections::HashSet;
+use std::path::{Path, PathBuf};
+use std::process::Command;
+
+/// Returns the set of additional libraries that need to be bundled with
+/// the given library, scanned recursively.
+///
+/// Any libraries in `provided_libs_paths` will be treated as available, without
+/// being emitted. Any other library not in `search_paths` or `provided_libs_paths`
+/// will result in an error.
+pub fn list_needed_libs_recursively(
+    lib: &Path,
+    search_paths: &[&Path],
+    provided_libs_paths: &[&Path],
+) -> Result<HashSet<PathBuf>> {
+    // Create a view of all libraries that are available on Android
+    let mut provided = HashSet::new();
+    for path in provided_libs_paths {
+        for lib in find_libs_in_dir(path)? {
+            // libc++_shared is bundled with the NDK but not available on-device
+            if lib != "libc++_shared.so" {
+                provided.insert(lib);
+            }
+        }
+    }
+
+    let mut to_copy = HashSet::new();
+
+    let mut artifacts = vec![lib.to_path_buf()];
+    while let Some(artifact) = artifacts.pop() {
+        for need in list_needed_libs(&artifact)? {
+            // c++_shared is available in the NDK but not on-device.
+            // Must be bundled with the apk if used:
+            // https://developer.android.com/ndk/guides/cpp-support#libc
+            let search_paths = if need == "libc++_shared.so" {
+                provided_libs_paths
+            } else {
+                search_paths
+            };
+
+            if provided.insert(need.clone()) {
+                if let Some(path) = find_library_path(search_paths, &need)? {
+                    to_copy.insert(path.clone());
+                    artifacts.push(path);
+                } else {
+                    bail!("Shared library `{}` not found", need);
+                }
+            }
+        }
+    }
+
+    Ok(to_copy)
+}
+
+/// List all required shared libraries as per the dynamic section
+fn list_needed_libs(library_path: &Path) -> Result<HashSet<String>> {
+    let mut readelf = Command::new("llvm-readelf");
+    let output = readelf.arg("--needed-libs").arg(library_path).output()?;
+    ensure!(
+        output.status.success(),
+        "Failed to run `{:?}`: {}",
+        readelf,
+        output.status
+    );
+    let output = std::str::from_utf8(&output.stdout).unwrap();
+    // TODO: Switch to `trim` because of CRLF on Windows?
+    let output = output.strip_prefix("NeededLibraries [\n").unwrap();
+    let output = output.strip_suffix("]\n").unwrap();
```

Comment:
Still need to test this...

Reply:
trimming it first can't hurt

Reply:
It's gonna be a bunch of `.trim()` calls, but sure.

Reply:
Fwiw `fn lines()` seems to be implemented as `split_terminator('\n')` which doesn't take `\r\n` into account?